autoload/phpactor/project/bufferMatcher.vim

```diff
@@ -0,0 +1,129 @@
+function! phpactor#project#bufferMatcher#assignFileToProject(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.matchFileToProject(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.getRepository().hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.getRepository().addProject(b:project)
+endfunction
+
+function! phpactor#project#bufferMatcher#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'getRepository': function('s:getRepository'),
+        \ 'matchFileToProject': function('s:matchFileToProject')
+        \ }
+endfunction
+
+function s:matchFileToProject(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
+    return l:project
+  endif
+
+  " early version to refactor or change
+  " heredoc are available since vim 8.1.1354
+  let l:message = printf(join([
+        \ 'Phpactor [RPC]',
```

Comment:
Is this related only to the RPC ?
If so I would propose to add it to the namespace so it's clear event by looking at the filsystem.
`autoload/phpactor/rpc/project/bufferMatcher.vim` maybe ?

Reply:
Yes, it is related only to used with RPC.  The whole phpactor code written in viml seems to be also related to RPC only ;) 

Most of this mechanism is as much phpactor independent as posible. It makes it easier to test and attach to or remove from phpactor. Maybe it would support root detection in any LSP client but I still have no knowledge how similar problems are resolved on both sides of LSP.

Reply:
>  The whole phpactor code written in viml seems to be also related to RPC only ;)
:D That's a damn valid point !

Reply:
I think eventually people will be able to use either LSP or RPC and have the same feature set - will have to see what the trade offs and advanrages are in the long run. RPC is obviously much lighter and could eventually properly use the indexer too.